lib/hsa/mcwamp_hsa.cpp

```diff
@@ -4587,10 +4577,124 @@ void HSADispatch::overrideAcquireFenceIfNeeded()
+        if (localDims[i] > workgroup_max_dim[i]) {
+          std::stringstream msg;
+          msg << "The extent of the tile (" << localDims[i] 
+              << ") exceeds the device limit (" << workgroup_max_dim[i] << ").";
+          throw Kalmar::runtime_exception(msg.str().c_str(), -1);
+        }
+        workgroup_size[i] = localDims[i];
+      }
+    }
+    else {
+
+      constexpr unsigned int recommended_flat_workgroup_size = 64;
+
+      // user didn't specify a workgroup size
+      if (dims == 1) {
+        workgroup_size[0] = recommended_flat_workgroup_size;
+      }
+      else if (dims == 2) {
+
+        // compute the group size for the 1st dimension
+        for (unsigned int i = 1; ; i<<=1) {
+          if (i == recommended_flat_workgroup_size
+              || i >= globalDims[0]) {
+            workgroup_size[0] = i;
+            break;
+          }
+        }
+
+        // compute the group size for the 2nd dimension
+        for (unsigned int j = 1; ; j<<=1) {
+          unsigned int flat_group_size = workgroup_size[0] * j;
+          if (flat_group_size == recommended_flat_workgroup_size
+              || j >= globalDims[1]) {
+            workgroup_size[1] = j;
+            break;
+          }
+        }
+      }
+      else if (dims == 3) {
+
+        // compute the group size for the 1st dimension
+        for (unsigned int i = 1; ; i<<=1) {
+          if (i == recommended_flat_workgroup_size
+              || i >= globalDims[0]) {
+            workgroup_size[0] = i;
+            break;
+          }
+        }
+
+        // compute the group size for the 2nd dimension
+        for (unsigned int j = 1; ; j<<=1) {
+          unsigned int flat_group_size = workgroup_size[0] * j;
+          if (flat_group_size == recommended_flat_workgroup_size
+              || j >= globalDims[1]) {
+            workgroup_size[1] = j;
+            break;
+          }
+        }
+
+        // compute the group size for the 3rd dimension
+        for (unsigned int k = 1; ; k<<=1) {
+          unsigned int flat_group_size = workgroup_size[0] * workgroup_size[1] * k;
+          if (flat_group_size == recommended_flat_workgroup_size
+              || k >= globalDims[2]) {
+            workgroup_size[2] = k;
+            break;
+          }
+        }
+      }
+    }
+
+    auto kernel = this->kernel;
+
+    auto calculate_kernel_max_flat_workgroup_size = [&] {
+      constexpr unsigned int max_num_vgprs_per_work_item = 256;
+      constexpr unsigned int num_work_items_per_simd = 64;
+      constexpr unsigned int num_simds_per_cu = 4;
+      const unsigned int workitem_vgpr_count = std::max((unsigned int)kernel->workitem_vgpr_count, 1u);
+      unsigned int max_flat_group_size = (max_num_vgprs_per_work_item / workitem_vgpr_count) 
+                                           * num_work_items_per_simd * num_simds_per_cu;
+      // round it to the closest wave size
+      max_flat_group_size -= max_flat_group_size%num_work_items_per_simd;
```

Comment:
I think since we multiplied by `num_work_items_per_simd`, then `max_flat_group_size` should already be a multiple of it. So we might not need to round to closest wave, unless this `(max_num_vgprs_per_work_item / workitem_vgpr_count)` produces a fraction, but we could guard it with a cast to unsigned int.

Reply:
makes sense, the rounding was removed in the newer version